Implement Hundredfaced Hapool Ja (Mamook)

CatsEyeXI opened the pull request:
<!-- Remove space and place 'x' mark between square [] brackets or click the checkbox after saving to affirm the following points: -->
<!-- (it should look like this: - [x] I have ...) -->
**_I affirm:_**
- [x] I have paid attention to this example and will edit again if need be to not break the formatting, or I will be ignored
- [x] I have read and understood the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md)
- [x] I've _**tested my code and the things my code has changed**_ since the last commit in the PR, and will test after any later commits

## What does this pull request do?

As it stands, on server init, the NM will be up along with his 4 'clones' by default. However
according to FFXIClopedia, and confirmed by me research in retail, the clones spawn only when NM uses Utsusemi.

## Steps to test these changes

engage mob, wait until it uses utsusemi, and clones should spawn around him, rather than being pre-spawned

zach2good commented on scripts/zones/Mamook/Zone.lua:
```diff
@@ -8,6 +8,9 @@ local zone_object = {}
     xi.helm.initZone(zone, xi.helm.type.LOGGING)
+	-- Clones should not automatically spawn on server init
+	-- TODO: There's probably a better way to handle this in the db
+    for i = 1, 4 do DespawnMob(ID.mob.HUNDREDFACED_HAPOOL_JA + i) end
```
You can change the spawn flags (I think in the groups or spawn points) to have them not spawn by default

TeoTwawki replied:
spawnType, set them to scripted spawns.

zach2good commented on scripts/zones/Mamook/mobs/Hundredfaced_Hapool_Ja.lua:
```diff
@@ -0,0 +1,39 @@
+    local spellId = spell:getID()
+        -- Utsusemi: Ichi (3 clones)
+        if spellId == xi.magic.spell.UTSUSEMI_ICHI then
+            for i = ID.mob.HUNDREDFACED_HAPOOL_JA + 1, ID.mob.HUNDREDFACED_HAPOOL_JA + 3 do
```
You don't need to use this id here, you can get it with mob:getID and then remove the new entry from the zone lua. Then it'll be shift proof

TeoTwawki replied:
```lua
local spellId = spell:getID()
local hundredfaced_hapool_ja = mob:getID()
-- Utsusemi: Ichi (3 clones)
if spellId == xi.magic.spell.UTSUSEMI_ICHI then
    for i = hundredfaced_hapool_ja + 1, hundredfaced_hapool_ja + 3 do
```
so that its only called once while being shift-proof per Zach's suggestion

CatsEyeXI replied:
Thanks for all the feedback. The only other issue I can't seem to move past, is making the clones update enmity to the target. (Seems like target is being lost on the new mob spawn) But I'll push a commit a bit later with the suggestions proposed and we can go from there.

TeoTwawki replied:
we'd need to get target from the main NM rather than relying on input the function arg, because in the context of this cast the target is itself (its where the spell is being cast onto)

`local hateTarget = GetMobByID(hundredfaced_hapool_ja):getTarget()`

`SpawnMob(i):updateEnmity(hateTarget)`

zach2good replied:
I did some similar "target-finding" here: <https://github.com/LandSandBoat/server/blob/base/modules/custom/lua/mazurka_generates_enmity.lua#L24-L29>

zach2good replied:
Teo fed you a very quick example here, but keep an eye on styling: <https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md#lua-misc--naming>

TeoTwawki commented on scripts/zones/Mamook/mobs/Hundredfaced_Hapool_Ja.lua:
```diff
@@ -0,0 +1,39 @@
+local entity = {}
+
+entity.onMobFight = function(mob, target)
+    mob:addListener("MAGIC_USE", "SPAWN_CLONES", function(mobArg, target, spell, action)
```
the listener should be added at spawn time instead of on every tick of onMobFight

CatsEyeXI commented:
Everything relating to the mob's actions appears to be correct now. 
My issue is converting this to a scripted spawn now. I've updated the spawntype variable to 128 - "Scripted", and added the logic to onMobDespawn... but now the mob won't pop unless done manually... what am i missing?

CatsEyeXI commented:
everything is working in this current state, however i did find the need for the static ID entry. 
if this is not correct or if any additional changes required let me know. thanks!

https://media.tenor.com/re_Txs4OUToAAAAd/tab-tab-key.gif

zach2good commented:
We're trying to move away from hard-coded IDs, since shifts are a major pain for us, downstream, and players.
You can get the mob-object for a mob by asking the zone for it, you can see me doing it here for FSA:
https://github.com/LandSandBoat/server/blob/base/scripts/zones/Batallia_Downs/Zone.lua#L32-L44

and here in a module:
<https://github.com/LandSandBoat/server/blob/base/modules/custom/lua/conquest_regional_npcs_always_up.lua>

It might look something like this:

```lua
zone_object.onInitialize = function(zone)
    xi.helm.initZone(zone, xi.helm.type.LOGGING)

    local hundredFacedHapoolJaID = zone:queryEntitiesByName("Hundredfaced_Hapool_Ja")[1]:getID()
    UpdateNMSpawnPoint(hundredFacedHapoolJaID)
    SpawnMob(hundredFacedHapoolJaID)
end
```

This is also the idea behind `DYNAMIC_LOOKUP`. This works especially well for NMs because they tend to have unique names in a zone that we don't have to disambiguate for any reason.

TeoTwawki commented:
> Everything relating to the mob's actions appears to be working properly now. My issue is converting this to a scripted spawn now. I've updated the spawntype value to 128 - "Scripted", and added the logic to onMobDespawn... but now the mob won't pop unless done manually... what am i missing?

His helpers spawned by his shadows would be type scripted, he himself would not be since he's a timed pop. If initial pop is randomized (like HNM timers tend to be for example) we'd init his time in the zone lua and then re-random every time his corpse despawns. The clones aren't the same ID and shouldn't have the same group so that they can be different spawn type and diff loot pool than the main NM.

zach2good commented:
Once this PR lands: https://github.com/LandSandBoat/server/pull/2882
it'll be safe for you to just use the table entry:`         HUNDREDFACED_HAPOOL_JA   = DYNAMIC_LOOKUP,`

CatsEyeXI commented:
Oooh. Very fancy. I'll give it a shot.

CatsEyeXI commented:
everything tested working with dynamic lookup, anything need to be done regarding these debug messages?
i believe these only print when there is no entry preset in nm_spawn_points... 

![image](https://user-images.githubusercontent.com/65316697/194129203-db1eca14-a290-4c5b-8be7-0df69d551a89.png)

cocosolos commented on scripts/zones/Mamook/mobs/Hundredfaced_Hapool_Ja.lua:
```diff
@@ -0,0 +1,64 @@
+    local hundredfaced_hapool_ja = mob:getID()
+    for i = 1, 4 do DespawnMob(hundredfaced_hapool_ja + i) end
+    -- TODO: 72 hours taken from DB, need retail verification.
+    UpdateNMSpawnPoint(hundredfaced_hapool_ja)
+    mob:setRespawnTime(259200)
```
Since there's nothing special about his spawn mechanics, you can just rely on the respawn time set in mob_groups and remove this and the changes to zone.lua and IDs.lua. Since it's a normal spawntype and has a respawn time set in the DB, all of this is handled automatically.

Commit:
Implement Hundredfaced Hapool Ja
